### PR TITLE
[GEOS-9331] Upgrade Freemarker to 2.3.31

### PR DESCRIPTION
**Pull request by aaime**
Upgrades the freemarker library to the latest version. The planted a "bomb" in the 2.3.21 version that was difficult to figure out, a change in behavior that was meant to be an improvement, but turned out to be very damaging in GeoServer. Thankfully, once figured out the nature of the issue, it was not difficult to work around it.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] ~~New unit tests have been added covering the changes~~ As a version upgrade, only checked that existing tests do pass, and updated them according to changed behavior (small changes in json escaping)
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

**Review by dromagnoli — APPROVED** [review]

**Review by bradh — COMMENTED** [review]
Should we set a legacy "fixed" version, as described in https://freemarker.apache.org/docs/pgui_config_settings.html ? I can't see that in the current code, so it could be a follow-on activity.

**Comment by aaime** [comment]
PR updated to resolve deprecated API usage. Also comes with fixing the reference version to the one that was already implied (basically replaced deprecated API calls by in-lining them, which reveals the defaults).

**Comment by aaime** [comment]
Also centralized the reference version number in a shared constant.

File: src/main/src/main/java/org/geoserver/template/TemplateUtils.java [hunk 1]
```diff
@@ -18,6 +21,9 @@
  * @author Kevin Smith, Boundless
  */
 public class TemplateUtils {
+    /** Reference feature version for Freemarker templates */
+    public static Version GS_VERSION = Configuration.VERSION_2_3_0;
```
Comment (bradh) [inline src/main/src/main/java/org/geoserver/template/TemplateUtils.java:25]: Its in the noise, but GS_VERSION seems a bit strange. FM_VERSION or FREEMARKER_CONFIG_VERSION?

Comment (aaime) [reply src/main/src/main/java/org/geoserver/template/TemplateUtils.java]: FM_VERSION it is, updated.

**Comment by bradh** [comment]
We have a bunch of stuff like https://github.com/geoserver/geoserver/blob/master/src/community/ogcapi/ogcapi-features/src/main/resources/org/geoserver/ogcapi/features/landingPage.ftl#L41

Is this change backwards-compatible (enough) that we don't need to deal with that yet?

**Comment by aaime** [comment]
@bradh those comments are reminders to improve the templates once we upgrade, so they are actually waiting for it, no backwards compatibility concern there (it's about doing stuff better once we have the newer version).

The upgrade will likely be backwards incompatible to some, the amount of changed piled up by Freemarker in these "bugfix releases" is nothing short of mind boggling, have a look, if you're more patient than me you might even read them all:
* https://freemarker.apache.org/docs/versions_2_3_19.html
* https://freemarker.apache.org/docs/versions_2_3_20.html
* https://freemarker.apache.org/docs/versions_2_3_21.html
* https://freemarker.apache.org/docs/versions_2_3_22.html
* https://freemarker.apache.org/docs/versions_2_3_23.html
* https://freemarker.apache.org/docs/versions_2_3_24.html
* https://freemarker.apache.org/docs/versions_2_3_25.html
* https://freemarker.apache.org/docs/versions_2_3_26.html
* https://freemarker.apache.org/docs/versions_2_3_27.html
* https://freemarker.apache.org/docs/versions_2_3_28.html
* https://freemarker.apache.org/docs/versions_2_3_29.html
* https://freemarker.apache.org/docs/versions_2_3_30.html
* https://freemarker.apache.org/docs/versions_2_3_31.html

That said, Freemarker has piled up a number of important upgrades, including fixing some CVEs. I'm proposing this PR only for master only, and we'll add a warning in the release notes. For 2.18.x we'll have to find another way.

**Pull request merged**